### PR TITLE
Config framework implementation

### PR DESCRIPTION
**Pull request by isra-fel**
This PR implements the abstractions defined in https://github.com/Azure/azure-powershell-common/pull/313

You can review this PR by commits:

1. 8efa0f96148d79a08ab04ac3a5eb35b06e077b6d brought in the code of Microsoft.Extensions.Configuration library we use as the storage layer, with some modification that allows us to update the configuration; to read environment variables from different targets (user/system) etc.
2. e54b3cbec5b7088174620f26b70d15996bd379d2 is the main implementation of the interfaces, the core types being `ConfigManager` which provides all the CRUD APIs to the config; and `ConfigInitializer` which initializes the config file and config manager. This commit also includes helper classes and abstract config definitions.
3. 3f59938167f26f711ce37a768eeb02b0b976d7ca is about the PowerShell cmdlets, for example, `Get-AzConfig`.
4. c1e967f8c7c465c6215031faea471824c1bd8ec3 is tests.

## Checklist

- [x] I have read the [_Submitting Changes_](../blob/master/CONTRIBUTING.md#submitting-changes) section of [`CONTRIBUTING.md`](../blob/master/CONTRIBUTING.md)
- [x] The title of the PR is clear and informative
- [x] The appropriate `ChangeLog.md` file(s) has been updated:
    - For any service, the `ChangeLog.md` file can be found at `src/{{SERVICE}}/{{SERVICE}}/ChangeLog.md`
    - A snippet outlining the change(s) made in the PR should be written under the `## Upcoming Release` header -- no new version header should be added
- [x] The PR does not introduce [breaking changes](../blob/master/documentation/breaking-changes/breaking-changes-definition.md)
- [x] If applicable, the changes made in the PR have proper test coverage
- [x] For public API changes to cmdlets:
    - [x] a cmdlet design review was approved for the changes in [this repository](https://github.com/Azure/azure-powershell-cmdlet-review-pr) (_Microsoft internal only_)
        - https://github.com/Azure/azure-powershell-cmdlet-review-pr/issues/1109
    - [x] the markdown help files have been regenerated using the commands listed [here](../blob/master/documentation/development-docs/help-generation.md#updating-all-markdown-files-in-a-module)

### CODE REVIEW

File: src/Accounts/Authentication/Config/ConfigInitializer.cs [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+
+using Microsoft.Azure.Commands.Common.Authentication.Abstractions;
+using Microsoft.Azure.Commands.Common.Authentication.Config.Internal.Interfaces;
+using Microsoft.Azure.PowerShell.Common.Config;
+using System;
+using System.Diagnostics;
+
+namespace Microsoft.Azure.Commands.Common.Authentication.Config
+{
+    public class ConfigInitializer
```
Comment (isra-fel) [inline src/Accounts/Authentication/Config/ConfigInitializer.cs:23]: doc

File: src/Accounts/Authentication/Config/ConfigInitializer.cs [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+            {
+                ValidateConfigFile();
+            }
+            return new ConfigManager(EnvironmentVariableProvider, DataStore, PathToConfigFile);
+        }
+
+        private void ValidateConfigFileContent()
+        {
+            string json = DataStore.ReadFileAsText(PathToConfigFile);
+            // todo: what to validate?
```
Comment (isra-fel) [inline src/Accounts/Authentication/Config/ConfigInitializer.cs:51]: todo

File: src/Accounts/Authentication/Config/ConfigInitializer.cs [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+            if (string.IsNullOrEmpty(json) || !json.TrimStart().StartsWith("{") || !json.TrimEnd().EndsWith("}"))
+            {
+                Debug.WriteLine($"The config file at {PathToConfigFile} could not be loaded.");
+                ResetConfigFileToDefault();
+            }
+
+            // todo: how to handle corrupted file?
```
Comment (isra-fel) [inline src/Accounts/Authentication/Config/ConfigInitializer.cs:58]: todo

File: src/Accounts/Authentication/Config/ConfigInitializer.cs [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+            //configManager.RegisterConfig(new SimpleTypedConfig<bool>(ConfigKeys.EnableInterceptSurvey, "When enabled, a message of taking part in the survey about the user experience of Azure PowerShell will prompt at low frequency.", true, "Azure_PS_Intercept_Survey"));
+            // todo: when the input is not a valid subscription name or id. Connect-AzAccount will throw an error. Is it right?
+            //configManager.RegisterConfig(new SimpleTypedConfig<string>(ConfigKeys.DefaultSubscriptionForLogin, "Subscription name or GUID. If defined, when logging in Azure PowerShell without specifying the subscription, this one will be used to select the default context.", string.Empty));
+            // todo: add later
+            //configManager.RegisterConfig(new RetryConfig());
+            // todo: how to migrate old config
+        //configManager.RegisterConfig(new EnableDataCollectionConfig());
+        }
+
+        private class RetryConfig : TypedConfig<int>
```
Comment (isra-fel) [inline src/Accounts/Authentication/Config/ConfigInitializer.cs:113]: remove this

File: src/Accounts/AuthenticationAssemblyLoadContext/AuthenticationAssemblyLoadContext.csproj [hunk 5]
```diff
@@ -3,7 +3,7 @@
     <PsModuleName>Accounts</PsModuleName>
   </PropertyGroup>
 
-  <Import Project="$(MSBuildThisFileDirectory)..\..\Az.props" />
+  <!--Import Project="$(MSBuildThisFileDirectory)..\..\Az.props" /-->
```
Comment (isra-fel) [inline src/Accounts/AuthenticationAssemblyLoadContext/AuthenticationAssemblyLoadContext.csproj:6]: Uncomment this until furthur validation

File: src/Accounts/Authentication/Config/Internal/ConfigurationRoot.cs [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+        {
+            get
+            {
+                return GetValueWithProviderId(key).Item1;
+            }
+            set
+            {
+                if (!_providers.Any())
+                {
+                    throw new InvalidOperationException("todo: SR.Error_NoSources");
```
Comment (isra-fel) [inline src/Accounts/Authentication/Config/Internal/ConfigurationRoot.cs:67]: Resolve the todos in internal folder

File: src/Accounts/Accounts/help/Clear-AzConfig.md [hunk 7]
```diff
@@ -0,0 +1,251 @@
...
+
+### ClearByKey
+```
+Clear-AzConfig [-PassThru] [-AppliesTo <String>] [-Scope <ConfigScope>]
+ [-DefaultProfile <IAzureContextContainer>] [-WhatIf] [-Confirm] [-DefaultSubscriptionForLogin]
+ [-EnableDataCollection] [-EnableInterceptSurvey] [-SuppressWarningMessage] [<CommonParameters>]
+```
+
+## DESCRIPTION
+{{ Fill in the Description }}
```
Comment (isra-fel) [inline src/Accounts/Accounts/help/Clear-AzConfig.md:29]: Docs will be updated in a following PR.
I merely included them for CI.

File: src/Accounts/Authentication/Config/ConfigInitializer.cs [hunk 8]
```diff
@@ -0,0 +1,156 @@
...
+                DataStore.WriteFile(_pathToConfigFile, @"{}");
+            }
+            catch (Exception ex)
+            {
+                // do not halt for IO exception
+                Debug.WriteLine(ex.Message);
+            }
+        }
+
+        // todo: tests initializes configs in a different way. Maybe there should be an abstraction IConfigInitializer and two concrete classes ConfigInitializer + TestConfigInitializer
```
Comment (isra-fel) [inline src/Accounts/Authentication/Config/ConfigInitializer.cs:133]: This refactoring is low priority

File: src/Accounts/Authentication/Config/ConfigInitializer.cs [hunk 9]
```diff
@@ -0,0 +1,156 @@
...
+        internal void InitializeForAzureSession(AzureSession session)
+        {
+            IConfigManager configManager = GetConfigManager();
+            session.RegisterComponent(nameof(IConfigManager), () => configManager);
+            RegisterConfigs(configManager);
+            configManager.BuildConfig();
+        }
+
+        private void RegisterConfigs(IConfigManager configManager)
```
Comment (isra-fel) [inline src/Accounts/Authentication/Config/ConfigInitializer.cs:142]: This method will be implemented in a follow up PR

File: src/Accounts/Accounts/Config/ConfigCommandBase.cs [hunk 10]
```diff
@@ -0,0 +1,98 @@
...
+using Microsoft.Azure.PowerShell.Common.Config;
+using Microsoft.WindowsAzure.Commands.Common.CustomAttributes;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Management.Automation;
+
+namespace Microsoft.Azure.Commands.Common.Authentication.Config
+{
+    [CmdletPreview("The cmdlet group \"AzConfig\" is in preview. Feedback is welcome: https://github.com/Azure/azure-powershell/discussions")]
```
Comment (isra-fel) [inline src/Accounts/Accounts/Config/ConfigCommandBase.cs:26]: We need a short link for this? Already have one?

Comment (isra-fel) [reply src/Accounts/Accounts/Config/ConfigCommandBase.cs:26]: https://aka.ms/azpsissue

File: src/Accounts/Authentication/Config/Helper/AppliesToHelper.cs [hunk 11]
```diff
@@ -0,0 +1,99 @@
...
+
+namespace Microsoft.Azure.Commands.Common.Authentication.Config
+{
+    /// <summary>
+    /// Helper class to deal with AppliesTo (how large is the scope that the config affects Azure PowerShell).
+    /// </summary>
+    public static class AppliesToHelper
+    {
+        internal static readonly Regex ModulePattern = new Regex(@"^az\.[a-z]+$", RegexOptions.IgnoreCase);
+        internal static readonly Regex CmdletPattern = new Regex(@"^[a-z]+-[a-z]+$", RegexOptions.IgnoreCase);
```
Comment (dingmeng-xue) [inline src/Accounts/Authentication/Config/Helper/AppliesToHelper.cs:30]: Cmdlet may contains digital number. such as `Get-AzDataFactoryV2`

File: src/Accounts/Authentication/Config/ConfigInitializer.cs (hunk 9 above)
Comment (dingmeng-xue) [reply src/Accounts/Authentication/Config/ConfigInitializer.cs:142]: It is great that those code can be removed.

File: src/Accounts/Authentication/Authentication.csproj [hunk 12]
```diff
@@ -30,4 +30,8 @@
     </EmbeddedResource>
   </ItemGroup>
 
+  <ItemGroup>
```
Comment (dingmeng-xue) [inline src/Accounts/Authentication/Authentication.csproj:33]: Can we remove this part? it seems it is not required.

File: src/Accounts/Authentication/Config/Helper/JsonConfigWriter.cs [hunk 13]
```diff
@@ -0,0 +1,153 @@
...
+            else
+            {
+                prop.Value = IsMultiContent(value) ? new JArray(value) : JToken.FromObject(value);
+            }
+        });
+
+        private bool IsMultiContent(object value)
+        {
+            return value is Array;
+        }
```
Comment (dingmeng-xue) [inline src/Accounts/Authentication/Config/Helper/JsonConfigWriter.cs:61]: seems not required. nice to have.

**Comment by dingmeng-xue** [comment]
A general question. If class is defined as interface, why is its namespace internal?

File: src/Accounts/Authentication/Authentication.csproj (hunk 12 above)
Comment (isra-fel) [reply src/Accounts/Authentication/Authentication.csproj:33]: removed

File: src/Accounts/Authentication/Config/Helper/JsonConfigWriter.cs (hunk 13 above)
Comment (isra-fel) [reply src/Accounts/Authentication/Config/Helper/JsonConfigWriter.cs:61]: Agreed, nice to have.

File: src/Accounts/Authentication/Config/Helper/AppliesToHelper.cs (hunk 11 above)
Comment (isra-fel) [reply src/Accounts/Authentication/Config/Helper/AppliesToHelper.cs:30]: Added numbers to the regex and moved the logic of handling regex into a new `PSNamingUtilities` class.

File: src/Accounts/Authentication/Config/ConfigInitializer.cs (hunk 9 above)
Comment (isra-fel) [reply src/Accounts/Authentication/Config/ConfigInitializer.cs:142]: removed

**Comment by isra-fel** [comment]
> A general question. If class is defined as interface, why is its namespace internal?

The code in the internal namespace works as the data layer of the config framework. It is not supposed to be exposed as public API, so I made it explicit by the namespace.

**Pull request merged**